src/services/download/mod.rs

```diff
@@ -0,0 +1,171 @@
+}
+
+impl Downloader {
+    pub fn new(worker: RemoteThread) -> Self {
+        Self { worker }
+    }
+
+    /// Download a file from a source and store it on the local filesystem.
+    ///
+    /// This does not do any deduplication of requests, every requested file is
+    /// freshly downloaded.
+    ///
+    /// # Arguments
+    ///
+    /// `source` - The source to download from.
+    ///
+    /// `dest` - Pathname of filename to save the downloaded file into.  The
+    ///    file will be created if it does not exist and truncated if it does.
+    ///    On successful completion the file's contents will be the download
+    ///    result.  In case of any error the file's contents is considered
+    ///    garbage.
+    ///
+    /// # Return value
+    ///
+    /// On success returns `Some(dest)`, if the download failed e.g. due to an
+    /// HTTP 404, `None` is returned.  If there is an error during the
+    /// downloading [`DownloadError`] is returned.
+    ///
+    /// [`DownloadError`]: types/struct.DownloadError.html
+    pub fn download(
+        &self,
+        source: SourceFileId,
+        dest: PathBuf,
+    ) -> Box<dyn Future<Item = DownloadStatus, Error = DownloadError> + Send + 'static> {
+        let fut03 = self.worker.spawn(|| async move {
+            let src_desc = source.to_string();
+            match source {
+                SourceFileId::Sentry(source, loc) => {
+                    let stream = sentry::download_stream(source, &loc);
+                    download_stream_fut(src_desc, stream, dest).compat().await
+                }
+                SourceFileId::Http(source, loc) => {
+                    let stream = http::download_stream(source, &loc);
+                    download_stream_fut(src_desc, stream, dest).compat().await
+                }
+                SourceFileId::S3(source, loc) => {
+                    let stream = s3::download_stream(source, &loc);
+                    download_stream_fut(src_desc, stream, dest).compat().await
+                }
+                SourceFileId::Gcs(source, loc) => {
+                    let stream = gcs::download_stream(source, &loc);
+                    download_stream_fut(src_desc, stream, dest).compat().await
+                }
+                SourceFileId::Filesystem(source, loc) => {
+                    filesystem::download_source(source, loc, dest)
+                }
+            }
+        });
+        let fut01 = fut03
+            .map(|spawn_ret| spawn_ret.unwrap_or_else(|_| Err(DownloadErrorKind::Canceled.into())))
+            .boxed()
+            .compat();
+        Box::new(fut01)
+    }
+}
+
+/// Download the source from a streaming future.
+///
+/// These streaming futures are currently implemented per source type.
+fn download_stream_fut(
+    source_desc: String,
+    stream: Box<dyn Future<Item = Option<DownloadStream>, Error = DownloadError>>,
+    dest: PathBuf,
+) -> Box<dyn Future<Item = DownloadStatus, Error = DownloadError>> {
```

Comment:
maybe as an excercise for later, this function is a prime candidate to convert to async/await. You are using `.compat()` up above. And inside here, you actually box twice.

Anyway, lets do this as a followup ;-)

Reply:
Yes, the types with futures01 are terrible.  This double boxing is because there a match and both arms need to return the same type.  But one returns an `AndThen` type while the other returns an `FutureResult` type.  And boxing them makes them both the "same" type.

The second box is needed because otherwise it also returns an `AndThen` type, which again is the wrong type.